File: Saver/CorkTexture.swift
import AppKit

/// A procedurally generated cork surface.
///
/// Rendered once into an image and cached: cork needs thousands of granules to
/// read as cork rather than as noise, and drawing that many shapes every frame
/// would cost far more than the whole rest of the scene.
enum CorkTexture {
    private static var cache: NSImage?
    private static var cachedSize: NSSize = .zero

    static func purge() { cache = nil; cachedSize = .zero }

    static func image(size: NSSize) -> NSImage {
        if let hit = cache,
           abs(cachedSize.width - size.width) < 2,
           abs(cachedSize.height - size.height) < 2 { return hit }

        let W = max(64, size.width), H = max(64, size.height)
        let img = NSImage(size: NSSize(width: W, height: H))
        img.lockFocus()

        let rect = NSRect(x: 0, y: 0, width: W, height: H)
        NSGradient(starting: NSColor(red: 0.71, green: 0.55, blue: 0.33, alpha: 1),
                   ending: NSColor(red: 0.56, green: 0.41, blue: 0.23, alpha: 1))?
            .draw(in: NSBezierPath(rect: rect), angle: -80)

        // Broad tonal mottling. Real cork is blotchy at a much larger scale than
        // its granules, and without this it reads as flat regardless of detail.
        for _ in 0..<70 {
            let r = CGFloat.random(in: min(W, H) * 0.07 ... min(W, H) * 0.30)
            let c = NSPoint(x: .random(in: 0...W), y: .random(in: 0...H))
            let light = Bool.random()
            let tone = light
                ? NSColor(red: 0.85, green: 0.70, blue: 0.46, alpha: 1)
                : NSColor(red: 0.36, green: 0.24, blue: 0.12, alpha: 1)
            NSGradient(colors: [tone.withAlphaComponent(light ? 0.10 : 0.13),
                                tone.withAlphaComponent(0)])?
                .draw(fromCenter: c, radius: 0, toCenter: c, radius: r, options: [])
        }

        // Granules: irregular, varied in tone, at cork's actual density.
        let granules = Int(W * H / 420)
        for _ in 0..<granules {
            let w = CGFloat.random(in: 2.5...11)
            let h = w * CGFloat.random(in: 0.45...1.6)
            let x = CGFloat.random(in: -w...W)
            let y = CGFloat.random(in: -h...H)
            let t = CGFloat.random(in: 0...1)
            let colour = NSColor(red: 0.42 + t * 0.42,
                                 green: 0.30 + t * 0.36,
                                 blue: 0.15 + t * 0.26,
                                 alpha: CGFloat.random(in: 0.10...0.34))
            colour.setFill()
            NSBezierPath(ovalIn: NSRect(x: x, y: y, width: w, height: h)).fill()
        }

        // Pits and pale flecks — the high-contrast specks the eye uses to judge
        // that a surface is granular rather than printed.
        for _ in 0..<(granules / 7) {
            let r = CGFloat.random(in: 1...3.4)
            NSColor(red: 0.20, green: 0.13, blue: 0.06,
                    alpha: CGFloat.random(in: 0.25...0.55)).setFill()
            NSBezierPath(ovalIn: NSRect(x: .random(in: 0...W), y: .random(in: 0...H),
                                        width: r, height: r * .random(in: 0.7...1.4))).fill()
        }
        for _ in 0..<(granules / 9) {
            let r = CGFloat.random(in: 1...3)
            NSColor(red: 0.93, green: 0.83, blue: 0.62,
                    alpha: CGFloat.random(in: 0.14...0.36)).setFill()
            NSBezierPath(ovalIn: NSRect(x: .random(in: 0...W), y: .random(in: 0...H),
                                        width: r, height: r)).fill()
        }

        img.unlockFocus()
        cache = img
        cachedSize = NSSize(width: W, height: H)
        return img
    }
}

File: windows/src/AlbumCoverScreenSaver.Saver/CorkTexture.cs
using AlbumCoverScreenSaver.Shared;
using SkiaSharp;

namespace AlbumCoverScreenSaver.Saver;

/// <summary>
/// A cork board, drawn once into a bitmap and kept.
/// </summary>
/// <remarks>
/// <para>
/// Cork needs thousands of granules to read as cork rather than as noise. At a
/// normal screen size that is about five thousand ellipses, plus seven hundred
/// pits and five hundred flecks, and drawing that every frame would cost more
/// than the whole rest of the scene put together. So it is generated once and
/// cached, and only regenerated when the surface actually changes size.
/// </para>
/// <para>
/// The layer order matters, because each one is composited over the last: a
/// base wash, broad mottling, granules, pits, then pale flecks. The mottling is
/// the layer people forget, and without it cork reads as flat no matter how much
/// fine detail is piled on top. The pits and flecks are the high-contrast specks
/// an eye uses to decide a surface is granular rather than printed.
/// </para>
/// </remarks>
internal sealed class CorkTexture : IDisposable
{
    private SKBitmap? _cached;
    private float _cachedWidth;
    private float _cachedHeight;

    /// <summary>
    /// The board at this size, generating it if the cached one is the wrong
    /// shape.
    /// </summary>
    /// <remarks>
    /// Two points of tolerance on each dimension, so a window that reports a
    /// sub-pixel difference between frames does not send five thousand ellipses
    /// through a redraw.
    /// </remarks>
    public SKBitmap? Get(float width, float height, Random random)
    {
        if (_cached is not null &&
            Math.Abs(_cachedWidth - width) < 2f &&
            Math.Abs(_cachedHeight - height) < 2f)
        {
            return _cached;
        }

        try
        {
            _cached?.Dispose();
            _cached = Generate(Math.Max(64, (int)width), Math.Max(64, (int)height), random);
            _cachedWidth = width;
            _cachedHeight = height;
        }
        catch (Exception error)
        {
            Log.Failure("generating the cork texture", error);
            _cached = null;
        }

        return _cached;
    }

    private static SKBitmap Generate(int width, int height, Random random)
    {
        var info = new SKImageInfo(width, height, SKColorType.Bgra8888, SKAlphaType.Premul);
        var board = new SKBitmap(info);

        using var canvas = new SKCanvas(board);
        var whole = SKRect.Create(0, 0, width, height);

        float Between(double low, double high) => (float)(low + (random.NextDouble() * (high - low)));

        // 1. The base. A near-vertical gradient, light at the top.
        using (var paint = new SKPaint { IsAntialias = true })
        {
            paint.Shader = SKShader.CreateLinearGradient(
                new SKPoint(width * 0.09f, 0f),
                new SKPoint(width * 0.91f, height),
                new[] { Colour(0.71f, 0.55f, 0.33f), Colour(0.56f, 0.41f, 0.23f) },
                SKShaderTileMode.Clamp);

            canvas.DrawRect(whole, paint);
            paint.Shader?.Dispose();
        }

        // 2. Broad tonal mottling. Real cork is blotchy at a far larger scale
        // than its granules, and without this the board reads as flat however
        // much fine detail goes on top of it.
        var shortest = Math.Min(width, height);
        for (var i = 0; i < 70; i++)
        {
            var radius = Between(shortest * 0.07, shortest * 0.30);
            var centre = new SKPoint(Between(0, width), Between(0, height));
            var light = random.Next(2) == 0;

            var tone = light ? Colour(0.85f, 0.70f, 0.46f) : Colour(0.36f, 0.24f, 0.12f);
            var alpha = light ? 0.10f : 0.13f;

            using var paint = new SKPaint { IsAntialias = true };
            paint.Shader = SKShader.CreateRadialGradient(
                centre, radius,
                new[] { tone.WithAlpha((byte)(alpha * 255f)), tone.WithAlpha(0) },
                SKShaderTileMode.Clamp);

            canvas.DrawCircle(centre, radius, paint);
            paint.Shader?.Dispose();
        }

        // 3. Granules. Hung off the left and bottom edges rather than stopping
        // short of them, or the board has a visible clean border.
        var granules = width * height / 420;
        using (var paint = new SKPaint { IsAntialias = true })
        {
            for (var i = 0; i < granules; i++)
            {
                var w = Between(2.5, 11.0);
                var h = w * Between(0.45, 1.6);
                var x = Between(-w, width);
                var y = Between(-h, height);

                var t = (float)random.NextDouble();
                paint.Color = Colour(0.42f + (0.42f * t), 0.30f + (0.36f * t), 0.15f + (0.26f * t))
                    .WithAlpha((byte)(Between(0.10, 0.34) * 255f));

                canvas.DrawOval(SKRect.Create(x, y, w, h), paint);
            }
        }

        // 4. Pits, and 5. pale flecks. These are the high-contrast specks an eye
        // uses to judge that a surface is granular rather than printed.
        using (var paint = new SKPaint { IsAntialias = true })
        {
            for (var i = 0; i < granules / 7; i++)
            {
                var r = Between(1.0, 3.4);
                paint.Color = Colour(0.20f, 0.13f, 0.06f)
                    .WithAlpha((byte)(Between(0.25, 0.55) * 255f));

                canvas.DrawOval(SKRect.Create(Between(0, width), Between(0, height), r, r * Between(0.7, 1.4)), paint);
            }

            for (var i = 0; i < granules / 9; i++)
            {
                var r = Between(1.0, 3.0);
                paint.Color = Colour(0.93f, 0.83f, 0.62f)
                    .WithAlpha((byte)(Between(0.14, 0.36) * 255f));

                canvas.DrawOval(SKRect.Create(Between(0, width), Between(0, height), r, r), paint);
            }
        }

        return board;
    }

    private static SKColor Colour(float r, float g, float b) => new(
        (byte)Math.Clamp(r * 255f, 0f, 255f),
        (byte)Math.Clamp(g * 255f, 0f, 255f),
        (byte)Math.Clamp(b * 255f, 0f, 255f));

    public void Purge()
    {
        _cached?.Dispose();
        _cached = null;
        _cachedWidth = 0;
        _cachedHeight = 0;
    }

    public void Dispose() => Purge();
}
